src/exec.cpp

```diff
@@ -996,6 +998,32 @@ void exec_job(parser_t &parser, job_t *j) {
         return;
     }
 
+    // Unfortunately `exec_job()` is called recursively when functions are encountered, with a new
+    // job id (and therefore pgrp) each time, but always from the main thread. This breaks terminal
+    // control since new pgrps take terminal control away from commands upstream in a different pgrp.
+    // We try to work around this with a heuristic to determine whether to reuse the same pgrp as the
+    // last-spawned pgrp if part of an existing job pipeline (keeping in mind that new jobs are
+    // recursively started for both foreground and background jobs, and that a function can expand
+    // to more than one external command, one (or more?) of which may want to read from upstream or
+    // write to downstream of a pipe.
+    // By keeping track of (select) "jobs in flight" we can try to marshall newly-created external
+    // processes into existing pgrps.
+    static std::stack<job_t *> active_jobs{};
```

Comment:
This entire thing strikes me as a bit of a hack, to be honest.

(The "real" solution is probably to pass the job along in exec_job, but that then means passing it all through the parser and that's just annoying)

I _think_ it works (and it seems to in actual tests), though.

Is there any way that a job in active_jobs becomes stale?

Reply:
It absolutely is a hack. But the entire parser approach is to set up a new context for the evaluation of functions and I broke too many things when I tried to change that.

`active_jobs` only contains `job_t` instances that haven't yet left `exec_job`, to keep track of the parent in event of recursive `exec_job` calls due to evaluation of functions and blocks. Jobs are removed from `active_jobs` before `exec_job` ends, so they're guaranteed to always be still "in execution" and very much valid.

I *can* change `exec_job` to take a `shared_ptr<job_t>` since that's what we actually originally have then decompose it to to a raw pointer for some reason, but there's really no need, it's guaranteed safe.

Reply:
I added a HACK warning and explanation of the preferred route plus complications you'll encounter going that way.